subsys/bluetooth/audio/has.c

```diff
@@ -34,7 +56,362 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
+
+	if (preset != NULL) {
+		preset->index = index;
+		preset->properties = properties;
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+		utf8_lcpy(preset->name, name, ARRAY_SIZE(preset->name));
+		preset->name[strlen(name)] = '\0';
+#else
+		preset->name = name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+	}
+
+	return preset;
+}
+
+static void preset_free(struct bt_has_preset *preset)
+{
+	preset->index = BT_HAS_PRESET_INDEX_NONE;
+}
+
+static void control_point_ntf_complete(struct bt_conn *conn, void *user_data)
+{
+	struct has_client *client = client_get(conn);
+
+	BT_DBG("conn %p\n", (void *)conn);
+
+	/* Resubmit if needed */
+	if (read_presets_req_is_pending(client)) {
+		k_work_submit(&client->control_point_work);
+	}
+}
+
+static void control_point_ind_complete(struct bt_conn *conn,
+				       struct bt_gatt_indicate_params *params,
+				       uint8_t err)
+{
+	if (err) {
+		/* TODO: Handle error somehow */
+		BT_ERR("conn %p err 0x%02x\n", (void *)conn, err);
+	}
+
+	control_point_ntf_complete(conn, NULL);
+}
+
+static int control_point_send(struct has_client *client, struct net_buf_simple *buf)
+{
+#if defined(CONFIG_BT_EATT)
+	/* TODO: Check whether EATT is connected */
+	if (bt_gatt_is_subscribed(client->conn, &has_svc.attrs[4], BT_GATT_CCC_NOTIFY)) {
+		client->params.ntf.attr = &has_svc.attrs[4];
+		client->params.ntf.func = control_point_ntf_complete;
+		client->params.ntf.data = buf->data;
+		client->params.ntf.len = buf->len;
+
+		return bt_gatt_notify_cb(client->conn, &client->params.ntf);
+	}
+#endif /* CONFIG_BT_EATT */
+
+	if (bt_gatt_is_subscribed(client->conn, &has_svc.attrs[4], BT_GATT_CCC_INDICATE)) {
+		client->params.ind.attr = &has_svc.attrs[4];
+		client->params.ind.func = control_point_ind_complete;
+		client->params.ind.destroy = NULL;
+		client->params.ind.data = buf->data;
+		client->params.ind.len = buf->len;
+
+		return bt_gatt_indicate(client->conn, &client->params.ind);
+	}
+
+	return -ECANCELED;
+}
+
+static int bt_has_cp_read_preset_rsp(struct has_client *client, struct bt_has_preset *preset,
+				     uint8_t is_last)
+{
+	struct bt_has_cp_hdr *hdr;
+	struct bt_has_cp_read_preset_rsp *rsp;
+
+	NET_BUF_SIMPLE_DEFINE(buf, sizeof(*hdr) + sizeof(*rsp) + BT_HAS_PRESET_NAME_MAX);
+
+	BT_DBG("conn %p preset %p is_last 0x%02x", (void *)client->conn, preset, is_last);
+
+	hdr = net_buf_simple_add(&buf, sizeof(*hdr));
+	hdr->opcode = BT_HAS_OP_READ_PRESET_RSP;
+	rsp = net_buf_simple_add(&buf, sizeof(*rsp));
+	rsp->is_last = is_last;
+	rsp->index = preset->index;
+	rsp->properties = preset->properties;
+	net_buf_simple_add_mem(&buf, preset->name, strlen(preset->name));
+
+	return control_point_send(client, &buf);
+}
+
+static uint8_t preset_find_next(const struct bt_has_preset *preset, void *user_data)
+{
+	const struct bt_has_preset **found = user_data;
+
+	if (*found == NULL || (*found)->index > preset->index) {
+		*found = preset;
+	}
```

Comment:
Is it missing a stop, or is continuing intentionally? 

```suggestion
		*found = preset;
		
		return BT_HAS_PRESET_ITER_STOP;
	}
```

Reply:
It's because the list entries are not sorted, so I loop over whole list. But I think I'll use quick sort just like you did in your CSIS PR, so that the presets will be sorted by ascending index.